capabilities/lib/zarf.ts

```diff
@@ -0,0 +1,81 @@
+import { ZarfState } from "../zarf-types";
+import { K8s, kind, Log } from "pepr";
+import {
+  zarfNamespace,
+  zarfStateSecret,
+  zarfImagePullSecret,
+  zarfManagedByLabel,
+} from "./constants";
+
+export async function getZarfRegistryURL(): Promise<string> {
+  try {
+    const secret = await K8s(kind.Secret)
+      .InNamespace(zarfNamespace)
+      .Get(zarfStateSecret);
+    const secretString = atob(secret.data!.state);
+    const zarfState: ZarfState = JSON.parse(secretString);
+    return zarfState.registryInfo.address;
+  } catch (err) {
+    throw new Error(
+      `unable to get registry URL from the ${zarfStateSecret} secret: ${JSON.stringify(
+        err,
+      )}`,
+    );
+  }
+}
+
+export async function updateZarfManagedImageSecrets(
+  ecrURL: string,
+  authToken: string,
+): Promise<void> {
+  try {
+    const namespace = await K8s(kind.Namespace).Get();
+    const namespaces = namespace.items;
+
+    for (const ns of namespaces) {
+      try {
+        await K8s(kind.Secret)
+          .InNamespace(ns.metadata!.name!)
+          .WithLabel(zarfManagedByLabel, "zarf")
+          .Get(zarfImagePullSecret);
+      } catch (err) {
+        // Continue checking the next namespace if this namespace doesn't have a "private-registry" secret
+        if (JSON.stringify(err).includes("404")) {
+          continue;
+        }
+        throw new Error(JSON.stringify(err));
+      }
```

Comment:
the `WithLabel` filter should happen during `K8s(kind.Namespace)`, as only namespaces managed by Zarf will have a Zarf secret contained. The 404 check is not needed if this is fixed.

Might be also worth making a PR to pepr to extend the type of `InNamespace` args to accept a ns object as the arg and do the `ns.metadata!.name!` check there.

Reply:
> the `WithLabel` filter should happen during `K8s(kind.Namespace)`, as only namespaces managed by Zarf will have a Zarf secret contained. The 404 check is not needed if this is fixed.
> 

Great point, updated

Reply:
> Might be also worth making a PR to pepr to extend the type of `InNamespace` args to accept a ns object as the arg and do the `ns.metadata!.name!` check there.

Will bring this up at Pepr office hours tomorrow. If the Pepr team accepts the change and releases a new version, we can refactor in a separate PR when renovate updates Pepr

Reply:
Following up with some notes from today's Pepr sync on this topic:

- Passing in an entire namespace object isn't the norm with k8s client tools

- This particular case is an ism of the upstream typescript k8s client model. Unsure why, but the upstream namespace model allows the `.metadata` field to be `undefined` which is why we have to use a [non-null operator (!)](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#non-null-assertion-operator-postfix-) to tell typescript that we are certain that the `.metadata` field is not undefined. [The `.metadata` field is a required field for the namespace object](https://kubernetes.io/docs/concepts/overview/working-with-objects/#required-fields), so we can be sure it isn't null/undefined since it's impossible to create a namespace without the metadata field, or remove the `.metadata` field from an existing object

<img width="729" alt="Screenshot 2023-12-13 at 1 24 19 PM" src="https://github.com/defenseunicorns/zarf-init-aws/assets/87675701/6209e44c-d182-4d4c-9618-bdaa883c21d4">